edge-util/src/Microsoft.Azure.Devices.Edge.Util/TaskEx.cs

```diff
@@ -125,6 +125,38 @@ public static async Task<T> TimeoutAfter<T>(this Task<T> task, TimeSpan timeout)
                 cts.Cancel();
                 return await task;
             }
+        }
+
+        public static async Task<T> TimeoutAfter<T>(this Task<T> task, TimeSpan timeout, Exception ex)
+        {

```

Comment:
I wonder, if we go back to action to throw an exception, maybe these two versions can be merged with a default parameter. Like creating a static action like () => thrown new TimeoutException("Operation timed out"), and then use that as a default parameter. This way the old codes would use the default action doing the same thing, and out specific version could throw the new exception. The only problem is that for reference types we can only use null as a default value, so the code must check if the action parameter is null, and then use the default action. Like:

static Action DefaultTimeoutAction = () => throw new TimeoutException("Operation timed out");
...
public static async Task<T> TimeoutAfter<T>(this Task<T> task, TimeSpan timeout, Action action)
{
 ...
 if (action == null)
{
  DefaultTimeoutAction ();
}
else 
{
  action();
}

Reply:
Reverted the changes for using an action, as well as merged the versions together.